From e9bfce5c126139c677c75eef58386309eba1ee86 Mon Sep 17 00:00:00 2001
From: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Date: Wed, 22 Nov 2017 16:42:29 -0800
Subject: [PATCH 01/10] dma: xilinx: dpdma: Fix programming 64bit address

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

addr_ext register is configured incorrectly which is
corrupting descriptor extension

Signed-off-by: Satish Kumar Nagireddy <satishna@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 7659beb40ca87cfd4305e08acc7e31ea77f6962d)
---
 drivers/dma/xilinx/xilinx_dpdma.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_dpdma.c b/drivers/dma/xilinx/xilinx_dpdma.c
index ca923b7..94833d0 100644
--- a/drivers/dma/xilinx/xilinx_dpdma.c
+++ b/drivers/dma/xilinx/xilinx_dpdma.c
@@ -650,8 +650,8 @@ static void xilinx_dpdma_sw_desc_addr_32(struct xilinx_dpdma_sw_desc *sw_desc,
 xilinx_dpdma_sw_desc_next_64(struct xilinx_dpdma_sw_desc *sw_desc,
 			     struct xilinx_dpdma_sw_desc *next)
 {
-	sw_desc->hw.next_desc = (u32)next->phys;
-	sw_desc->hw.addr_ext |= ((u64)next->phys >> 32) &
+	sw_desc->hw.next_desc = lower_32_bits(next->phys);
+	sw_desc->hw.addr_ext |= upper_32_bits(next->phys) &
 				XILINX_DPDMA_DESC_ADDR_EXT_ADDR_MASK;
 }
 
@@ -671,10 +671,13 @@ static void xilinx_dpdma_sw_desc_addr_64(struct xilinx_dpdma_sw_desc *sw_desc,
 {
 	struct xilinx_dpdma_hw_desc *hw_desc = &sw_desc->hw;
 	unsigned int i;
+	u32 src_addr_extn;
 
-	hw_desc->src_addr = (u32)dma_addr[0];
-	hw_desc->addr_ext |=
-		((u64)dma_addr[0] >> 32) & XILINX_DPDMA_DESC_ADDR_EXT_ADDR_MASK;
+	hw_desc->src_addr = lower_32_bits(dma_addr[0]);
+	src_addr_extn = upper_32_bits(dma_addr[0]) &
+			XILINX_DPDMA_DESC_ADDR_EXT_ADDR_MASK;
+	hw_desc->addr_ext |= (src_addr_extn <<
+			      XILINX_DPDMA_DESC_ADDR_EXT_ADDR_SHIFT);
 
 	if (prev)
 		xilinx_dpdma_sw_desc_next_64(prev, sw_desc);
-- 
1.8.5.2.233.g932f7e4

